src/lib/snarky_js_bindings/lib/snarky_js_bindings_lib.ml

```diff
@@ -2856,10 +2864,21 @@ module Ledger = struct
     |> Mina_base.Account_id.public_key
     |> Signature_lib.Public_key.Compressed.to_string |> Js.string
 
-  let custom_token_id pk token =
+  let forest_digest_of_field_checked :
+      Field.t -> Parties.Digest.Forest.Checked.t =
+    Obj.magic
+
+  let custom_token_id_checked pk token =
+    Mina_base.Account_id.Checked.derive_token_id
+      ~owner:
+        (Mina_base.Account_id.Checked.create
+           (Signature_lib.Public_key.Compressed.var_of_t (public_key pk))
```

Comment:
using `var_of_t` is wrong because it's only possible if the variable has been converted to a value first, which is what the `public_key` function does. The forced converting to a value that `public_key` does only works inside the prover, or if the `pk` is a constant, but not when compiling the circuit.

We want to preserve the same underlying variables that `pk` contains (two field elements; possibly variables) and convert them to a `Public_key.Compressed.var` in an in-snark way. This is code that I got to type-check that I think achieves that:
```ocaml
  let public_key_checked (pk : public_key) :
      Signature_lib.Public_key.Compressed.var =
    let x = pk##.g##.x##.value in
    let y = pk##.g##.y##.value in
    Signature_lib.Public_key.compress_var (x, y) |> Impl.run_checked
```
(although I'm hazy on what `run_checked` means)

Reply:
Got it! I changed the implementation and removed the use of `var_of_t`. The code snippet type checks and seems to pass existing test cases :+1: